runtime/src/iree/hal/utils/event_pool.h

```diff
@@ -0,0 +1,95 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef IREE_HAL_UTILS_EVENT_POOL_H_
+#define IREE_HAL_UTILS_EVENT_POOL_H_
+
+#include "iree/base/api.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif  // __cplusplus
+
+//===----------------------------------------------------------------------===//
+// iree_hal_wrapped_event_t
+//===----------------------------------------------------------------------===//
+
+// Implementation type for iree_hal_event_t.
+// It maps to CUevent for CUDA and hipEvent_t for HIP.
+typedef struct iree_hal_event_impl_t* iree_hal_event_impl_t;
+
+// Symbol table for event related implementation APIs.
+typedef struct iree_hal_event_impl_symtable_t {
+  iree_status_t(IREE_API_PTR* create)(void* user_data,
+                                      iree_hal_event_impl_t* out_event);
+  void(IREE_API_PTR* destroy)(void* user_data, iree_hal_event_impl_t event);
+  iree_status_t(IREE_API_PTR* synchronize)(void* user_data,
+                                           iree_hal_event_impl_t event);
+} iree_hal_event_impl_symtable_t;
+
+// An struct that wraps a CUevent/hipEvent_t object with a reference count for
+// lifetime management.
+//
+// iree_hal_wrapped_event_t objects cannot be directly created; they should be
+// acquired from the event pool and released back to the pool once done.
+//
+// Thread-safe; multiple threads may retain and release the same event.
+typedef struct iree_hal_wrapped_event_t iree_hal_wrapped_event_t;
+
+// Returns the underlying CUevent/hipEvent_t handle behind |event|.
+iree_hal_event_impl_t iree_hal_wrapped_event_handle(
+    const iree_hal_wrapped_event_t* event);
+
+// Retains the given |event| by increasing its reference count.
+void iree_hal_wrapped_event_retain(iree_hal_wrapped_event_t* event);
+
+// Releases the given |event| by decreasing its reference count.
+//
+// |event| will be returned to its owning pool when the reference count is 0.
+void iree_hal_wrapped_event_release(iree_hal_wrapped_event_t* event);
```

Comment:
Is this (`iree_hal_wrapped_event_t`) really needed? It seems redundant with `iree_hal_event_t`